File: api/src/modules/calendar/calendar.service.ts
import prisma from '../../lib/prismaClient';
import { AppError } from '../../middleware/errorHandler';
import { EventType } from '@prisma/client';

interface CreateEventData {
  title: string;
  description?: string;
  type: EventType;
  scheduledAt: Date;
  duration?: number;
  recurrence?: any;
  notificationTime?: number;
}

interface UpdateEventData extends Partial<CreateEventData> {}

class CalendarService {
  async getEvents(userId: number, page: number = 1, limit: number = 20) {
    const skip = (page - 1) * limit;

    const [events, total] = await Promise.all([
      prisma.calendarEvent.findMany({
        where: { userId },
        skip,
        take: limit,
        orderBy: { scheduledAt: 'asc' },
      }),
      prisma.calendarEvent.count({ where: { userId } }),
    ]);

    return {
      events,
      pagination: { page, limit, total, totalPages: Math.ceil(total / limit) },
    };
  }

  async getEvent(userId: number, eventId: number) {
    const event = await prisma.calendarEvent.findFirst({
      where: { id: eventId, userId },
    });

    if (!event) {
      throw new AppError('Event not found', 404);
    }

    return event;
  }

  async createEvent(userId: number, data: CreateEventData) {
    const event = await prisma.calendarEvent.create({
      data: {
        ...data,
        userId,
      },
    });

    return event;
  }

  async updateEvent(userId: number, eventId: number, data: UpdateEventData) {
    await this.getEvent(userId, eventId);

    const event = await prisma.calendarEvent.update({
      where: { id: eventId },
      data,
    });

    return event;
  }

  async deleteEvent(userId: number, eventId: number) {
    await this.getEvent(userId, eventId);

    await prisma.calendarEvent.delete({
      where: { id: eventId },
    });

    return { message: 'Event deleted successfully' };
  }

  async completeEvent(userId: number, eventId: number) {
    await this.getEvent(userId, eventId);

    const event = await prisma.calendarEvent.update({
      where: { id: eventId },
      data: {
        isCompleted: true,
        completedAt: new Date(),
      },
    });

    return event;
  }

  async getUpcoming(userId: number, limit: number = 10) {
    const events = await prisma.calendarEvent.findMany({
      where: {
        userId,
        scheduledAt: { gte: new Date() },
        isCompleted: false,
      },
      take: limit,
      orderBy: { scheduledAt: 'asc' },
    });

    return events;
  }
}

export default new CalendarService();

File: mobile/lib/services/calendar_service.dart
import '../models/calendar_event.dart';
import 'api_client.dart';

class CalendarService {
  final ApiClient _client;

  CalendarService(this._client);

  Future<List<CalendarEvent>> getEvents({
    DateTime? startDate,
    DateTime? endDate,
    String? type,
    int page = 1,
    int limit = 50,
  }) async {
    final params = {
      'page': page.toString(),
      'limit': limit.toString(),
    };
    
    if (startDate != null) {
      params['startDate'] = startDate.toIso8601String();
    }
    
    if (endDate != null) {
      params['endDate'] = endDate.toIso8601String();
    }
    
    if (type != null) {
      params['type'] = type;
    }

    final json = await _client.get('/calendar/events', params: params);
    final data = json['data'] as Map<String, dynamic>;
    final eventsData = data['events'] as List;
    
    return eventsData.map((e) => CalendarEvent.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<CalendarEvent> getEvent(int eventId) async {
    final json = await _client.get('/calendar/events/$eventId');
    final data = json['data'] as Map<String, dynamic>;
    return CalendarEvent.fromJson(data);
  }

  Future<CalendarEvent> createEvent({
    required String title,
    String? description,
    required String type,
    required DateTime scheduledAt,
    int? duration,
    Map<String, dynamic>? recurrence,
    int? notificationTime,
  }) async {
    final json = await _client.post('/calendar/events', body: {
      'title': title,
      'description': description,
      'type': type,
      'scheduledAt': scheduledAt.toIso8601String(),
      'duration': duration,
      'recurrence': recurrence,
      'notificationTime': notificationTime,
    });

    final data = json['data'] as Map<String, dynamic>;
    return CalendarEvent.fromJson(data);
  }

  Future<CalendarEvent> updateEvent({
    required int eventId,
    String? title,
    String? description,
    String? type,
    DateTime? scheduledAt,
    int? duration,
    bool? isCompleted,
    Map<String, dynamic>? recurrence,
    int? notificationTime,
  }) async {
    final json = await _client.put('/calendar/events/$eventId', body: {
      if (title != null) 'title': title,
      if (description != null) 'description': description,
      if (type != null) 'type': type,
      if (scheduledAt != null) 'scheduledAt': scheduledAt.toIso8601String(),
      if (duration != null) 'duration': duration,
      if (isCompleted != null) 'isCompleted': isCompleted,
      if (recurrence != null) 'recurrence': recurrence,
      if (notificationTime != null) 'notificationTime': notificationTime,
    });

    final data = json['data'] as Map<String, dynamic>;
    return CalendarEvent.fromJson(data);
  }

  Future<void> deleteEvent(int eventId) async {
    await _client.delete('/calendar/events/$eventId');
  }

  Future<void> completeEvent(int eventId) async {
    await _client.post('/calendar/events/$eventId/complete');
  }
}


